File: Initial_Modulations/main_compare_modulations.m
clear; clc; close all;

% Recursively add all subfolders in the current directory
addpath(genpath(pwd));

% Creates a configuration structure (cfg)
cfg.seed          = 42;
cfg.numBits       = 2e6;      % keep divisible by 2, 3, 4, 6 if possible
cfg.EbNoVec       = 0:1:15;
cfg.plotEbNo      = 6;        % noisy constellation snapshot
cfg.showClean     = true;
cfg.showNoisy     = true;

% Initialize random number generator (rng) using the Mersenne Twister algo
rng(cfg.seed, 'twister');

specs = [
    struct('name', "BPSK",   'type', "psk",    'M', 2,  'phaseOffset', 0)
    struct('name', "QPSK",   'type', "psk",    'M', 4,  'phaseOffset', pi/4)
    struct('name', "8-PSK",  'type', "psk",    'M', 8,  'phaseOffset', pi/8)
    struct('name', "8-QAM",  'type', "genqam", 'M', 8,  'phaseOffset', 0)
    struct('name', "16-QAM", 'type', "qam",    'M', 16, 'phaseOffset', 0)
    struct('name', "64-QAM", 'type', "qam",    'M', 64, 'phaseOffset', 0)
];

% Preallocate a cell array to store the simulation results for each scheme
results = cell(numel(specs), 1);    % dimension: len(specs) x 1

% Loops through each defined modulation scheme
for i = 1:numel(specs)
    fprintf('Running %s...\n', specs(i).name);
    results{i} = run_modulation_case(specs(i), cfg);

    if cfg.showClean
        figure('Name', specs(i).name + " Clean Constellation");
        plot_constellation(results{i}.txSig, specs(i).name + " (No Noise)", results{i}.refConstellation);
    end

    if cfg.showNoisy && ~isempty(results{i}.rxSigPlot)
        figure('Name', specs(i).name + " Noisy Constellation");
        plot_constellation(results{i}.rxSigPlot, ...
            specs(i).name + " - AWGN at " + string(cfg.plotEbNo) + " dB", ...
            results{i}.refConstellation);
    end
end

figure('Name', 'BER Comparison');
plot_ber_curve(cfg.EbNoVec, results);

save('results/phase_1_results.mat', 'cfg', 'specs', 'results');
disp('Done. Saved to phase1_results.mat');

File: Phase1_Modulations/main_compare_modulations.py
import os
import numpy as np
import matplotlib.pyplot as plt
from run_modulation_case import run_modulation_case


def main():
    # Configuration structure for the simulation
    cfg = {
        "seed": 42,
        "num_bits": int(1e5),
        "EbNoVec": np.arange(0, 21, 1),
        "plotEbNo": 12,
        "showClean": True,
        "showNoisy": True,
    }

    np.random.seed(cfg["seed"])

    specs = [
        {"name": "BPSK", "type": "psk", "M": 2, "phaseOffset": 0},
        {"name": "QPSK", "type": "psk", "M": 4, "phaseOffset": np.pi / 4},
        {"name": "8-PSK", "type": "psk", "M": 8, "phaseOffset": np.pi / 8},
        {"name": "8-QAM", "type": "genqam", "M": 8, "phaseOffset": 0},
        {"name": "16-QAM", "type": "qam", "M": 16, "phaseOffset": 0},
        {"name": "64-QAM", "type": "qam", "M": 64, "phaseOffset": 0},
    ]

    results = []
    for spec in specs:
        print(f"Running {spec['name']}...")
        result = run_modulation_case(spec, cfg)
        results.append(result)

        if cfg["showClean"]:
            plt.figure(figsize=(6, 6))
            import plotting.plot_constellation as pc

            pc.plot_constellation(result["txSig"], f"{spec['name']} (No Noise)", result["refConstellation"])
            plt.title(f"{spec['name']} Clean Constellation")
            plt.show()

        if cfg["showNoisy"] and result["rxSigPlot"] is not None:
            plt.figure(figsize=(6, 6))
            import plotting.plot_constellation as pc

            pc.plot_constellation(result["rxSigPlot"], f"{spec['name']} - AWGN at {cfg['plotEbNo']} dB", result["refConstellation"])
            plt.title(f"{spec['name']} Noisy Constellation")
            plt.show()

    import plotting.plot_ber_curve as pbc

    plt.figure(figsize=(8, 6))
    pbc.plot_ber_curve(cfg["EbNoVec"], results)
    plt.title("BER Comparison")
    plt.show()

    os.makedirs("results", exist_ok=True)
    np.savez(
        "results/phase_1_results.npz",
        cfg=cfg,
        specs=specs,
        results=results,
    )
    print("Done. Saved to results/phase_1_results.npz")


if __name__ == "__main__":
    main()
